Preparing captions section to support multiple captions

LeonardoRosaa opened the pull request:
## What has changed and why?

Preparing the captions section to support multiple captions. This PR adds scrolling when the captions don’t fit within the available space.

## How has it been tested?


https://github.com/user-attachments/assets/2df55d98-ebdf-4322-8d0e-0ade4f765868



## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

michal-lightly commented on lightly_studio/src/lightly_studio/api/routes/api/caption.py:
```diff
@@ -9,21 +9,23 @@
-) -> GetAllCaptionsResult:
+) -> GetAllCaptionsFromSampleResult:
     """Retrieve captions for a dataset."""
-    return caption_resolver.get_all(
```
This is the only place we use `caption_resolver.get_all`. We can remove the function (or better said - modify it to fit the new needs).

LeonardoRosaa commented:
This PR introduces a few additional changes:

- Cleaned up the caption view classes.
- Removed the `get_all `function from the Caption Resolver. 
- The `GET /captions `endpoint now uses `get_all_captions_by_samples` instead.

JonasWurst commented on lightly_studio/src/lightly_studio/resolvers/caption_resolver.py:
```diff
@@ -39,42 +40,55 @@ def create_many(session: Session, captions: Sequence[CaptionCreate]) -> list[Cap
+    count_subquery = (
+        select(SampleTable.sample_id)
+        .join(CaptionTable)
+        .where(CaptionTable.dataset_id == dataset_id)
```
see comment above

JonasWurst commented on lightly_studio/src/lightly_studio/resolvers/caption_resolver.py:
```diff
@@ -39,42 +40,55 @@ def create_many(session: Session, captions: Sequence[CaptionCreate]) -> list[Cap
+            col(CaptionTable.created_at).asc(),
+            col(CaptionTable.caption_id).asc(),
+        )
+        .distinct()
```
Is the distinct required here?

JonasWurst replied:
I would appreciate a comment, explaining the full query:

Getting all samples, that hold captions. Ordered by the caption creation date and their id

JonasWurst commented on lightly_studio/src/lightly_studio/resolvers/caption_resolver.py:
```diff
@@ -39,42 +40,55 @@ def create_many(session: Session, captions: Sequence[CaptionCreate]) -> list[Cap
 
 
-def get_all(
+def get_all_captions_by_sample(
```
The view used is not a caption view but rather a samples view. This is not obvious from the description.

Can we instead use the sampleView as is? Or if required generate a caption focused SampleView.

The problem I see: like this, we won't reuse it.

JonasWurst commented on lightly_studio/src/lightly_studio/resolvers/caption_resolver.py:
```diff
@@ -39,42 +40,55 @@ def create_many(session: Session, captions: Sequence[CaptionCreate]) -> list[Cap
     Returns:
-        List of captions matching the filters, total number of captions, next cursor (pagination)
+        List of captions matching the filters, total number of samples with captions, next
+        cursor (pagination)
```
This comment does not match the actual return type

JonasWurst commented on lightly_studio/tests/api/routes/api/test_caption.py:
```diff
@@ -64,15 +71,17 @@ def test_read_captions__last_page(
-    assert result["total_count"] == 4
+    assert result["total_count"] == 10
     assert result["nextCursor"] is None
-    assert [f"Caption number {i}" for i in range(2, 4)] == [data["text"] for data in result["data"]]
```
This test is not sufficient anymore, we also need to test the filename or sample_id. At the moment, we might receive captions from the wrong sample as they are all similar.

Hence, smth in this direction:
```
    assert result["data"][0].sample_id = # get. the sample_id here, this should be the 2-th sample id 
    assert result["data"][1].sample_id = # get. the sample_id here, this should be the 3-th sample id 
    ...
```

JonasWurst commented on lightly_studio/tests/conftest.py:
```diff
@@ -410,14 +410,14 @@ def captions_test_data(
-        )
-
-        captions_to_create.append(caption)
+    for sample in samples:
```
We can omit the inner loop, as there is no need any more to keep 4 captions. We needed them before, to test the pagination, now pagination is not happening over captions --> 2 is fine.

The outer loop, i feel is fine, since it is just used for test case creation.

JonasWurst commented on lightly_studio/tests/resolvers/test_caption_resolver.py:
Major: This file needs rework, after the changes, we are not testing the captions. we only access the number of samples.

JonasWurst commented on lightly_studio/src/lightly_studio/models/caption.py:
```diff
@@ -57,17 +49,18 @@ class CaptionView(SQLModel):
 
-class CaptionDetailsView(CaptionView):
-    """Response model for caption."""
+class CaptionViewsBySample(BaseModel):
```
This new name is not yet well suited. Let's discuss in more detail :)

JonasWurst reviewed (changes requested):
Thanks for the PR. This is the right direction, but we will need some changes.

In general, I agree that packing the captions by samples is a more elegant approach. To avoid confusion in the future, I would try to make this more clear. E.g. I think it is rather a sample view than a CaptionView.

We also need to rework the tests, after the current changes, some of them do not test anymore what they were intended for.

michal-lightly commented on lightly_studio/src/lightly_studio/models/caption.py:
```diff
@@ -57,17 +49,18 @@ class CaptionView(SQLModel):
     model_config = ConfigDict(populate_by_name=True)
 
-    captions: List[CaptionDetailsView] = PydanticField(..., alias="data")
+    samples: List[CaptionViewsBySample] = PydanticField(..., alias="data")
```
Where do we construct CaptionViewsBySample?

LeonardoRosaa replied:
We've the class declaration [here](https://github.com/lightly-ai/lightly-studio/blob/1d87e4bcfad752028fc7cc789327b11045a33bf5/lightly_studio/src/lightly_studio/models/caption.py#L52).

michal-lightly commented on lightly_studio/src/lightly_studio/resolvers/caption_resolver.py:
```diff
@@ -39,42 +40,57 @@ def create_many(session: Session, captions: Sequence[CaptionCreate]) -> list[Cap
-    query = select(CaptionTable).order_by(
-        col(CaptionTable.created_at).asc(),
-        col(CaptionTable.caption_id).asc(),
+    """Selects distinct samples with captions, and orders by creation time and caption ID."""
```
Minor: There is an issue with the comment formatting.

michal-lightly commented on lightly_studio/src/lightly_studio/resolvers/caption_resolver.py:
```diff
@@ -9,13 +9,14 @@
 
 
-class GetAllCaptionsResult(BaseModel):
+class GetAllCaptionsFromSampleResult(BaseModel):
```
This class has nothing to do with captions. Let me see the rest of the changes.

michal-lightly commented on lightly_studio/src/lightly_studio/models/caption.py:
```diff
@@ -57,17 +51,11 @@ class CaptionView(SQLModel):
-
-
-class CaptionsListView(BaseModel):
+class CaptionViewsBySampleWithCount(BaseModel):
```
This class has nothing to do with captions. It is confusing it is placed here.

LeonardoRosaa commented:
- Moved the logic to sample_resolver;
- Removed the old captions endpoint;
- Simplified the tests;
- Kept the order_by since it works;
- Didn’t remove the .subquery() because I received a few warnings from SQLModel.
